### ISSUE TITLE
hooks don't work on subclasses

### ISSUE THREAD

**Issue by stevecrozz** [jekyll/jekyll#3870]
Currently Jekyll triggers hooks by matching the class of the triggering instance against a map of classes to symbols. This causes a problem described in octopress/paginate#7 where hooks are not triggered because the instance in question is a subclass of Page and not an instance of Page.

We need to come up with a way to support this use case.

**Comment by imathis** [comment]
Thanks so much!

**Comment by parkr** [comment]
We should use `is_a?` in my opinion instead of `class` if possible. That may require a slight API change, though, yeah? Instead of `:document`, we'd pass `Jekyll::Document` (Symbol being replaced by a Class).

**Comment by stevecrozz** [comment]
@parkr: I think what it comes down to is what we want to do here:
https://github.com/jekyll/jekyll/blob/master/lib/jekyll/hooks.rb#L92

It would be a bigger change than just swapping `class` for `is_a?` because in order to do it we'd have to enumerate the options and call `is_a?` for each one. We can do it without even an external interface change, but I hesitate because Hash lookups are so much faster than calling is_a? in a loop.

We could complicate the internal trigger interface by requiring callers to pass the owner symbol in addition to their current parameters. That would help keep hide the complexity from end users and could also be done without changing the external interface. I think this is along the lines of what @envygeeks proposed to me, but I'd like to confirm with him.

**Comment by envygeeks** [comment]
We might have to take a double approach unless we want to force people to reintegrate.  Here is an example:

```ruby
Jekyll::Hooks.trigger :site, :post_render, data
Jekyll::Hooks.trigger data, :post_render
```

We are late in the 3.0 cycle but there is a need for hooks to be more flexible, there are many cases where we would need to highly reconfigure pieces of Jekyll's core to get hooks integrated into them and if we don't accept both we break one or the other way late in our release schedule.

Lets take for example my want to add a `:post_init` hook for `:configuration` so that plugins can attach themselves and create default configurations, or other various hook places.  Right now this can be complicated or overly loose because what we had assumed was a Configuration being passed ended up being a Hash being passed, but if we had the ability to trigger like the above we could solve this problem.

The decision has to be made to either make triggers generic allowing anybody to call or to make them based on class which simplifies your code, if it comes down to it, my vote is going to be accepting `Jekyll::Hooks.trigger :site, :post_render, data` because it ultimately simplifies everything and does not have any specific tie.

**Comment by stevecrozz** [comment]
There is no specific need to pass a magic object as the first parameter and use it to lookup the hooks other than it simplifies the internal trigger interface. That's not a good enough reason to cause grief for people who want to use triggers without tying them to a specific class.

So I'm on board with @envygeeks idea and I think I can put this together without causing any backwards incompatibilities as long as no external users are calling Jekyll::Hooks#trigger.

**Comment by stevecrozz** [comment]
The change seems pretty innocuous to me. Let me know what you guys think.

**Comment by stevecrozz** [comment]
fixed by #3871
